### ISSUE TITLE
[HLSL] Scalarizer incorrectly changes i64 vector calls to i32 calls

### ISSUE THREAD

**Issue by V-FEXrt** [llvm/llvm-project#144966]
### Notes from Chris:

https://godbolt.org/z/oj3o6dzqK

Caused by this PR: https://github.com/llvm/llvm-project/pull/111569

`AreAllVectorsOfMatchingSize` is probably true, which is causing it to pick the return type as the overload instead of the argument type

### Historic notes:

Issues was detected in https://github.com/llvm/offload-test-suite/pull/209 where the assertion below was raised

```
# | Assertion failed: (i >= FTy->getNumParams() || FTy->getParamType(i) == Args[i]->getType()) && "Calling a function with a bad signature!", file E:\actions-runner\_work\offload-test-suite\offload-test-suite\llvm-project\llvm\lib\IR\Instructions.cpp, line 757
# | PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace, preprocessed source, and associated run script.
# | Stack dump:
# | 0.	Program arguments: E:\\actions-runner\\_work\\offload-test-suite\\offload-test-suite\\llvm-project\\build\\bin\\clang-dxc.exe -cc1 -triple dxilv1.5-unknown-shadermodel6.5-compute -O3 -emit-obj -dumpdir a- -disable-free -clear-ast-before-backend -main-file-name source.hlsl -mrelocation-model static -mframe-pointer=all -fmath-errno -ffp-contract=on -fno-rounding-math -mconstructor-aliases -debugger-tuning=gdb -fdebug-compilation-dir=E:\\actions-runner\\_work\\offload-test-suite\\offload-test-suite\\llvm-project\\build\\tools\\OffloadTest\\test\\clang-d3d12\\Feature\\HLSLLib -fcoverage-compilation-dir=E:\\actions-runner\\_work\\offload-test-suite\\offload-test-suite\\llvm-project\\build\\tools\\OffloadTest\\test\\clang-d3d12\\Feature\\HLSLLib -resource-dir E:\\actions-runner\\_work\\offload-test-suite\\offload-test-suite\\llvm-project\\build\\lib\\clang\\21 -ferror-limit 19 -O3 -finclude-default-header -fgnuc-version=4.2.1 -fskip-odr-check-in-gmf -vectorize-loops -vectorize-slp -o C:\\WINDOWS\\SERVIC~1\\NETWOR~1\\AppData\\Local\\Temp\\lit-tmp-n1_2nql9\\source-cacd82.obj -x hlsl E:\\actions-runner\\_work\\offload-test-suite\\offload-test-suite\\llvm-project\\build\\tools\\OffloadTest\\test\\clang-d3d12\\Feature\\HLSLLib\\Output\\firstbitlow.64.test.tmp/source.hlsl
# | 1.	<eof> parser at end of file
# | 2.	Code generation
# | 3.	Running pass 'Function Pass Manager' on module 'E:\actions-runner\_work\offload-test-suite\offload-test-suite\llvm-project\build\tools\OffloadTest\test\clang-d3d12\Feature\HLSLLib\Output\firstbitlow.64.test.tmp/source.hlsl'.
# | 4.	Running pass 'Scalarize vector operations' on function '@main'
# | Exception Code: 0xC000001D
# |  #0 0x00007ff61e7831c6 HandleAbort E:\actions-runner\_work\offload-test-suite\offload-test-suite\llvm-project\llvm\lib\Support\Windows\Signals.inc:429:0
# |  #1 0x00007ffb3150[198](https://github.com/llvm/offload-test-suite/actions/runs/15766564388/job/44444055767?pr=209#step:10:199)9 (C:\WINDOWS\System32\ucrtbase.dll+0xc1989)
# |  #2 0x00007ffb314e4ab1 (C:\WINDOWS\System32\ucrtbase.dll+0xa4ab1)
# |  #3 0x00007ffb31502986 (C:\WINDOWS\System32\ucrtbase.dll+0xc2986)
# |  #4 0x00007ffb31502b61 (C:\WINDOWS\System32\ucrtbase.dll+0xc2b61)
# |  #5 0x00007ff61ec28947 llvm::CallInst::init(class llvm::FunctionType *, class llvm::Value *, class llvm::ArrayRef<class llvm::Value *>, class llvm::ArrayRef<class llvm::OperandBundleDefT<class llvm::Value *>>, class llvm::Twine const &) E:\actions-runner\_work\offload-test-suite\offload-test-suite\llvm-project\llvm\lib\IR\Instructions.cpp:755:0
# |  #6 0x00007ff61e47c33c llvm::CallInst::CallInst E:\actions-runner\_work\offload-test-suite\offload-test-suite\llvm-project\llvm\include\llvm\IR\Instructions.h:1680:0
# |  #7 0x00007ff61e47c33c llvm::CallInst::Create(class llvm::FunctionType *, class llvm::Value *, class llvm::ArrayRef<class llvm::Value *>, class llvm::ArrayRef<class llvm::OperandBundleDefT<class llvm::Value *>>, class llvm::Twine const &, class llvm::InsertPosition) E:\actions-runner\_work\offload-test-suite\offload-test-suite\llvm-project\llvm\include\llvm\IR\Instructions.h:1568:0
# |  #8 0x00007ff61e47bdf6 llvm::IRBuilderBase::CreateCall(class llvm::FunctionType *, class llvm::Value *, class llvm::ArrayRef<class llvm::Value *>, class llvm::Twine const &, class llvm::MDNode *) E:\actions-runner\_work\offload-test-suite\offload-test-suite\llvm-project\llvm\include\llvm\IR\IRBuilder.h:2507:0
# |  #9 0x00007ff61e71e3ce llvm::IRBuilderBase::CreateCall E:\actions-runner\_work\offload-test-suite\offload-test-suite\llvm-project\llvm\include\llvm\IR\IRBuilder.h:2528:0
# | #10 0x00007ff61e71e3ce `anonymous namespace'::ScalarizerVisitor::splitCall E:\actions-runner\_work\offload-test-suite\offload-test-suite\llvm-project\llvm\lib\Transforms\Scalar\Scalarizer.cpp:793:0
```

**Comment by farzonl** [comment]
This theory does not seem to be correct `AreAllVectorsOfMatchingSize` is false.
```gdb
(lldb) expr CI->dump()
  %hlsl.firstbitlow.i = tail call <4 x i32> @llvm.dx.firstbitlow.v4i16(<4 x i16> %4)
(lldb) n
Process 54801 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = step over
    frame #0: 0x000000010303c220 clang-dxc`(anonymous namespace)::ScalarizerVisitor::splitCall(this=0x000000016fdf4020, CI=0x000000012f0121c0) at Scalarizer.cpp:689:30
   686  bool ScalarizerVisitor::splitCall(CallInst &CI) {
   687    Type *CallType = CI.getType();
   688    bool AreAllVectorsOfMatchingSize = isStructOfMatchingFixedVectors(CallType);
-> 689    std::optional<VectorSplit> VS;
   690    if (AreAllVectorsOfMatchingSize)
   691      VS = getVectorSplit(CallType->getContainedType(0));
   692    else
Target 0: (clang-dxc) stopped.
(lldb) expr AreAllVectorsOfMatchingSize
(bool) $0 = false
(lldb) c
Process 54801 resuming
Assertion failed: ((i >= FTy->getNumParams() || FTy->getParamType(i) == Args[i]->getType()) && "Calling a function with a bad signature!"), function init, file Instructions.cpp, line 757.
Process 54801 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = hit program assert
    frame #4: 0x0000000101f748d8 clang-dxc`llvm::CallInst::init(this=0x000000011d204300, FTy=0x000000012800cae8, Func=0x000000011d204138, Args=ArrayRef<llvm::Value *> @ 0x000000016fdf2ea0, Bundles=ArrayRef<llvm::OperandBundleDefT<llvm::Value *> > @ 0x000000016fdf2e90, NameStr=0x000000016fdf31b8) at Instructions.cpp:755:5
   752           "Calling a function with bad signature!");
   753 
   754    for (unsigned i = 0; i != Args.size(); ++i)
-> 755      assert((i >= FTy->getNumParams() ||
   756              FTy->getParamType(i) == Args[i]->getType()) &&
   757             "Calling a function with a bad signature!");
   758  #endif
Target 0: (clang-dxc) stopped.
```

**Comment by farzonl** [comment]
This theory is incorrect. The PR linked did not introduce a bug. the i32s are because we never defined an overload for firstbitlow for the arguments. So whats happeining is the return type ended up being the only type on the `Tys` stack.
```cpp
if (isVectorIntrinsicWithOverloadTypeAtArg(ID, -1, TTI))
    Tys.push_back(VS->SplitTy);
```

We never get the argument type. Ie this line does not happen

```cpp
if (isVectorIntrinsicWithOverloadTypeAtArg(ID, I, TTI)) {
        OverloadIdx[I] = Tys.size();
        Tys.push_back(OpVS->SplitTy);
      }
```
The reason this is an issue for `firstbitlow` is because the return type of this intrinsic is allways a int32 while the argument type can be any integer size.
